src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala

```diff
@@ -404,49 +405,65 @@ object PhaseBlock extends LazyLogging {
     val blocks = ListBuffer[PhaseBlock]()
     logger.info(s"Found ${phaseBlocks.size} phase block")
-    phaseBlocks.values.toList.sortBy { b => (b.start, -b.length) } match {
-      case Nil                    => Unit
-      case block :: Nil           => blocks.append(block)
-      case block :: currentBlocks =>
-        var left = block
-        currentBlocks.foreach { right =>
-          require(left.getStart <= right.getStart)
-          if (left.overlaps(right)) {
-            if (!modifyBlocks) {
-              throw new IllegalStateException(s"Block $left overlaps $right")
-            }
-            if (left.encloses(right)) { // keep the left block only
-              logger.info(s"Removing $right enclosed in $left")
-              // do nothing
-            }
-            else if (right.encloses(left)) { // keep the right block
-              logger.info(s"Removing $left enclosed in $right")
-              left = right
-            }
-            else if (left.length < right.length) { // truncate left
-              logger.info(s"Truncating $left which overlaps $right");
-              blocks.append(left.copy(end=right.getStart-1))
-              left = right
-            }
-            else { // truncate right
-              logger.info(s"Truncating $right which overlaps $left");
-              blocks.append(left)
-              left = right.copy(start=left.getEnd+1)
-            }
-          }
-          else {
-            blocks.append(left)
-            left = right
-          }
+
+    var left = blockSet.pollFirst()
+    while (!blockSet.isEmpty) {
+      val right = blockSet.pollFirst()
+
+      if (right.getStart < left.getStart) {
+        throw new IllegalArgumentException(s"left: $left right: $right")
+      }
+      require(left.getStart <= right.getStart)
+      if (left.overlaps(right)) {
+        if (!modifyBlocks) {
+          throw new IllegalStateException(s"Block $left overlaps $right")
+        }
+        if (left.encloses(right)) { // keep the left block only
+          logger.info(s"Removing $right enclosed in $left")
+          // do nothing
+        }
+        else if (right.encloses(left)) { // keep the right block
+          logger.info(s"Removing $left enclosed in $right")
+          left = right
+        }
+        else if (left.length < right.length) { // truncate left
+          logger.info(s"Truncating $left which overlaps $right");
+          blocks.append(left.copy(end=right.getStart-1))
+          left = right
         }
-        // don't forget the last one
+        else { // truncate right
+          logger.info(s"Truncating $right which overlaps $left");
+          blocks.append(left)
+          // Since the start position of the right block is changed, and we are going to use it in the next iteration,
+          // then re-insert it into the set and poll.
+          blockSet.add(right.copy(start=left.getEnd+1))
+          left = blockSet.pollFirst()
```

Comment:
I see.  The only other suggestion I have for making this clearer would to rename `blockSet` to `blocksIn` and `blocks` to `blocksOut` since you push things back into both of them.